src/main/java/net/floodlightcontroller/forwarding/Forwarding.java

```diff
@@ -228,46 +200,661 @@ private void removeExpiredFlowSetId(U64 flowSetId, NodePortTuple avoid, Iterator
+     */
+    protected void doL3Routing(Ethernet eth, IOFSwitch sw, OFPacketIn pi, IRoutingDecision decision,
+                             FloodlightContext cntx, @Nonnull VirtualGatewayInstance gatewayInstance, OFPort inPort) {
+
+        MacAddress gatewayMac = gatewayInstance.getGatewayMac();
+
+        if (eth.getPayload() instanceof IPv4) {
+            IPv4Address intfIpAddress = findInterfaceIP(gatewayInstance, ((IPv4) eth.getPayload()).getDestinationAddress());
+            if (intfIpAddress == null) {
+                log.debug("Can not locate corresponding interface for gateway {}, check its interface configuration",
+                        gatewayInstance.getName());
+                return;
+            }
+        }
+
+        if (isBroadcastOrMulticast(eth)) {
+            // When cross-subnet, host send ARP request to gateway. Gateway need to generate ARP response to host
+            if (eth.getPayload() instanceof ARP && ((ARP) eth.getPayload()).getOpCode().equals(ARP.OP_REQUEST)
+                    && gatewayInstance.isAGatewayInft(((ARP) eth.getPayload()).getTargetProtocolAddress())) {
+                IPacket arpReply = gatewayArpReply(cntx, gatewayMac);
+                pushArpReply(arpReply, sw, OFBufferId.NO_BUFFER, OFPort.ANY, inPort);
+                log.debug("Virtual gateway pushing ARP reply message to source host");
+            }
+            else {
+                doFlood(sw, pi, decision, cntx);
+            }
+        }
+        else {
+            // This also includes L2 forwarding
+            doL3ForwardFlow(sw, pi, decision, cntx, gatewayInstance, false);
+        }
+
+    }
+
+
+    /**
+     * This function retrieves a patch from source device to destination device and then install flows(L3/L2) over the path.
+     * In L3 case, virtual router caches the packet-in while it sends the ARP message to learn destination device
+     *
+     * @param sw
+     * @param pi
+     * @param decision
+     * @param cntx
+     * @param gateway
+     * @param requestFlowRemovedNotifn
+     */
+    protected void doL3ForwardFlow(IOFSwitch sw, OFPacketIn pi, IRoutingDecision decision, FloodlightContext cntx,
+                                   VirtualGatewayInstance gateway, boolean requestFlowRemovedNotifn) {
+        Ethernet eth = IFloodlightProviderService.bcStore.get(cntx, IFloodlightProviderService.CONTEXT_PI_PAYLOAD);
+        OFPort srcPort = OFMessageUtils.getInPort(pi);
+
+        MacAddress virtualGatewayMac = gateway.getGatewayMac();
+        DatapathId srcSw = sw.getId();
+        IDevice dstDevice = IDeviceService.fcStore.get(cntx, IDeviceService.CONTEXT_DST_DEVICE);
+        IDevice srcDevice = IDeviceService.fcStore.get(cntx, IDeviceService.CONTEXT_SRC_DEVICE);
+
+        if (dstDevice == null) {
+            // Try one more time to retrieve dst device
+            if (eth.getPayload() instanceof IPv4 && eth.getDestinationMACAddress().equals(virtualGatewayMac)) {
```

Comment:
Would it be more intuitive to check the EthType? I.E. eth.getEtherType() == EtherType.IPv4